src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
+            {
+                return cachedTokens;
+            }
+
+            // 2) If the cache doesn't contain an exact match for the range, we can still check if there's a
+            // partial match. This allows us to reduce the range we have to compute tokens for. We want to find
+            // the cached range with the most overlap with the requested range.
+            (Range Range, int NumOverlappingLines)? currentBestMatch = null;
+            foreach (var rangeCandidate in documentCache.Keys)
+            {
+                if (!range.OverlapsWith(rangeCandidate))
+                {
+                    continue;
+                }
+
+                var overlap = range.Overlap(rangeCandidate);
+                Assumes.NotNull(overlap);
+
+                // It's possible that the cached range candidate encompasses a span that completely overlaps with
+                // the requested range. Nothing in the LSP spec explicitly prevents us from returning excess
+                // tokens outside of the requested range, so if we hit this case we can return early.
+                if (overlap.Equals(range) && documentCache.TryGetValue(rangeCandidate, out cachedTokens))
+                {
+                    return cachedTokens;
+                }
+
+                // The range candidate becomes the best range either if:
+                //     1) It is the first candidate we are analyzing
+                //     2) The # of overlapping lines exceeds the current best candidate, *and* the current candidate
+                //     is not in the middle of the requested range. (We avoid the latter since it complicates later
+                //     computations.)
+                var numOverlappingLines = overlap.End.Line - overlap.Start.Line;
+                if (currentBestMatch is null ||
+                    ((currentBestMatch.Value.NumOverlappingLines < numOverlappingLines) && (rangeCandidate.Start <= range.Start || rangeCandidate.End >= range.End)))
+                {
+                    currentBestMatch = (rangeCandidate, numOverlappingLines);
+                }
+            }
+
+            // We couldn't find a best range match. Don't use any cache values and compute tokens normally.
+            if (currentBestMatch is null)
+            {
+                return null;
+            }
+
+            if (!documentCache.TryGetValue(currentBestMatch.Value.Range, out cachedTokens))
+            {
+                logger.LogError("Could not find the expected cached tokens.");
+                return null;
+            }
+
+            // 3) We've found the best range match. There are now two paths to consider depending on whether
+            // we need to populate the start or end of the range.
+            var bestMatch = currentBestMatch.Value;
+
+            // 3a) The latter part of the range needs to be computed (i.e. the beginning is already cached).
+            if (bestMatch.Range.Start <= range.Start)
```

Comment:
I think we only need to recompute if it's less than start right? Similar question to the end `>=` below
```suggestion
            if (bestMatch.Range.Start < range.Start)
```